### DISCUSSION TITLE
Lint error "Imports of lazy-loaded libraries are forbidden"

### DISCUSSION THREAD

**Discussion by nzacca** [nrwl/nx#15441] [General]
Aloha @JamesHenry !

Sorry to trouble you but I found [this old comment](https://github.com/nrwl/nx/issues/4463#issuecomment-760075946) which greatly improved my understanding of how projects are linted (thank you) but I'm coming up short understanding why the following scenario is failing linting (Imports of lazy-loaded libraries are forbidden). On the surface, it feels like a bug but I'd like to understand why this is occurring.

We have the following structure:

- apps
  - foo
    - src
      - app
        - `lazy-import-of-foo-lib.ts`
          - i.e. `import("@example/foo-lib")`
    - testing
      - `static-import-of-foo-lib-testing.ts`  --> Linting error here on the static import
        - i.e. `import { foo } from "@example/foo-lib/testing"`
- libs
  - foo-lib
    - src
      - index.ts
    - testing
      - index.ts

With the following paths defined in tsconfig:

```json
paths: [
  "@example/foo-lib/testing": ["libs/foo-lib/testing/index.ts"], // NOT Nx library
  "@example/foo-lib": ["libs/foo-lib/src/index.ts"] // Nx library
]
```

Note that _only_ `foo-lib` is an Nx Project and `foo-lib/testing` is just a defined path.

Now, according to the issue linked above, Nx understands the relationships between Apps and Libs and not folders. I would have expected then that Nx would only look at the `sourceRoot` defined in `project.json` to determine which files should be considered when evaluating the static and dynamic imports for the lint rule. It seems though the rule is looking at the parent folder `libs/foo-lib`. This feels like a bug to me since Nx shouldn't know anything about the outside `testing` folder.

At first glance this scenario seems similar to the original issue linked above but it is slightly different. The `foo-lib/testing` files here are not within the `foo-lib/src` folder. If they were, I could understand why the linting rule is flagging this as an error. This rule used to pass in 15.7 but is now failing for us in 15.8.

Any insight would be greatly appreciated.

**Comment by JamesHenry** [comment]
`foo-lib/testing` still exists within the `foo-lib` project, though? TypeScript path mappings are used to resolve file paths, but it doesn't change that those file paths exist within a project.

To understand better what's going on in your particular case you have two good options:

- Try running `nx graph` and focusing on the two "nodes" (projects) in question. Click on the "edge" (the arrow) between them to see where Nx has tracked the relationships between them.

E.g. here you can see a static relationship was found based on the named file:

<img width="441" alt="image" src="https://user-images.githubusercontent.com/900523/223065412-1ab21a4b-f5d3-486c-92e0-27579cb5adbe.png">


- Remember the lint rule is just JavaScript in your `node_modules`, so at any point you can go into `eslint-plugin-nx` and step through what it is doing and what specific things it is resolving for your particular workspace.

---

Ultimately the lint rule is trying to protect you from a costly mistake - invalidating attempts to lazy-load code. You can resolve the lint error in one of two ways:
- Refactor things so that they is project-level separation between lazy-loaded and non lazy-loaded code
- In cases where you know it is definitely not applicable, and where refactoring is undesirable, you can always use an `eslint-disable` comment.

**Comment by nzacca** [comment]
@JamesHenry 

Thank you for the response. I followed your instructions but I'm still missing something. Here's the graph for the test scenario above:

![image](https://user-images.githubusercontent.com/871384/223190182-28bccd03-c81f-4c88-8722-e6497d091297.png)

Edit: Sorry about that. I was testing some more changes and made the testing lib a project which produced the image above but that did not get rid of the linting error. Removing the project does produce a branch from `foo` to both `foo-lib` and `foo-lib-testing`. But I'm still confused as to why the Nx would still produce a linting error even though the libraries are clearing distinct after making them their own projects.

~There is no relationship between `foo-lib-testing` and `foo-lib`. When you say "exists within the project" - what do you mean? Do you mean within the same folder? From an Nx perspective, there is no `foo-lib/testing` since there is is no Nx project. From a Typescript perspective, the two folders are isolated. The `src` folder and `testing` folder are only present in either `tsconfig.app` (src) or `tsconfig.spec` (testing) but they are never compiled together.~ 

Thanks again for your time.

**Reply by nzacca** [reply]
@JamesHenry 

I got it. Looks like the dependency graph was still cached with the old tree which is why it was still producing the linting error. Thank you so much for your time.
